Add support for core (cljs.test) unit tests running in PhantomJS

rorygibson opened the pull request:
David Nolen released the core test functionality in late December.
This PR adds support for placing tests using that syntax ("deftest" etc) into ```test/cljs``` and running them using PhantomJS.
Enabled using --cljstest (I've tried to follow the patterns laid out in the Speclj feature)

Comments appreciated!

plexus commented:
Hey @rorygibson , thanks for the patch! This is something that was much needed.

I think in this case it doesn't have to be behind a feature flag. Having tests is best practice, and clojurescript.test is as default as it gets at this point. Plus the amount of feature flags is making testing harder and harder, slowing down new releases.

Are you ok with changing things so the speclj flag toggles between clojurescript.test and speclj?

What do you think of also adding a clojure.test example and instructions?

rorygibson commented:
Yeah, good plan.
I'll update the PR, and I'll add some instructions / examples too.



On 29 January 2015 at 10:22, Arne Brasseur <notifications@github.com> wrote:

> Hey @rorygibson <https://github.com/rorygibson> , thanks for the patch!
> This is something that was much needed.
>
> I think in this case it doesn't have to be behind a feature flag. Having
> tests is best practice, and clojurescript.test is as default as it gets at
> this point. Plus the amount of feature flags is making testing harder and
> harder, slowing down new releases.
>
> Are you ok with changing things so the speclj flag toggles between
> clojurescript.test and speclj?
>
> What do you think of also adding a clojure.test example and instructions?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/plexus/chestnut/pull/87#issuecomment-72001866>.
>



-- 
*Rory Gibson*
Chief Technology Officer
Bright North

*t:* +44 20 3598 2217 *m:* +44 7714 719280 *w**:* brightnorth.co.uk

New Loom House, 101 Back Church Lane, London, E1 1LU
Company number: 08340170

 *Finalist* - Best Innovation, MRS Awards 2014
 *Finalist* - Best eCommerce Product, eCommerce Awards for Excellence 2014
 *Winner* - Client and Agency Integrated Data Planning Team, DataIQ Talent
Awards 2014
 *Finalist* - Data Scientist, DataIQ Talent Awards 2014


[image: Bright North]

Find out why optimising images improves conversion rates in our newest
whitepaper <http://bit.ly/1tKmBMH>.

Sign up to our newsletter <http://eepurl.com/8ea_L> for trending ecommerce
news and expert tips.

We are always looking for exceptional people to join our team. If you’re
super talented and want to work in a fast moving, evolutionary company, get
in touch <hello@brightnorth.co.uk?subject=Recruitment>.

rorygibson commented:
@plexus PR updated.
Now with 
* CLJS core test as default, 
* Speclj still available as an option, 
* an example Clojure test in test/clj, 
* some documentation (TESTING.md) 
* all the necessary tweaks to make these things hang together.

Please review, comments appreciated!

plexus commented:
love it :heart:  also super great you added documentation

Would you mind moving the documentation file to the doc site? Just put it under `doc/` and link to it from `doc/SUMMARY.md`. I'll push out a new gh-pages once this is in. I would link to that from the generated README (`src/leiningen/new/chestnut/README.md`). In general info on what to do with your app after `lein new` belongs there and in the docs. The project README is more for an overview and some very basic getting started info.

The only thing I'm unsure about is the layout of files under `resources/`. Seeing `private` there seems a bit strange to me... maybe `resources/dev/`? Not sure... maybe we should look for some examples in the wild of where people stick these clojurescript test support files. Either way fine to let it go in like this already.

Other small nitpicks

* Add an entry to the change log in the project README. Don't forget to give yourself credit :smile: 
* few whitespace fixes. Especially when it comes to the generated files I'm quite particular about this. Will comment those inline

Finally... I suppose you tested all this? :stuck_out_tongue:

rorygibson commented:
@plexus I've updated the PR again :-)

**Docs:**
* testing doc moved into gitbook format (it's the first time i've done that, please review in case I've gotten it wrong)
* trimmed down testing-related specifics in the README
* Added changelog entry - had to create a new stanza so chose today's date for the snapshot name 

**File locations**
I've moved the stuff out of ```resources/private``` and created a space under ```env/test``` which more closely matches how the rest of Chestnut fits together, I think.
This means I've also moved the polyfill into ```env/test/js```; I've updated the speclj script to reflect this.

This has the nice side-effect that ```resources/public/js``` is now empty until a build is done, so could now be set as a lein clean-target - which would mean cleaning works properly.
(I haven't implemented this, as it's possible that Chestnut users might be placing their own JS libraries into that folder and I don't want to delete them... Maybe think about generating compiled JS into ```resources/public/compiled-js``` or something in a future release?)

**Testing**
As far as testing goes: I've tried out multiple combinations of command line flags, created some projects using the template, run the tests & specs in them, checked that CLJX still works when enabled, etc etc.
I'll be really interested to see what you come up with for testing the template itself, right now it's difficult to be sure I've caught all possible edge cases.

plexus commented on README.md:
```diff
@@ -154,6 +156,10 @@ reports and pull requests are very welcome.
 
 ## Changelog
 
+### v0.7.0-SNAPSHOT-20150130
```
that should be master until we actually push a new snapshot to clojars

rorygibson replied:
Will change to "v0.7.0"

plexus replied:
Just `master` is fine, makes it clear these things aren't released yet.

plexus commented on doc/testing.md:
```diff
@@ -1,30 +0,0 @@
-# Testing
-
-(This page is about testing Chestnut itself, *not* about how to do testing in your Chestnut app.)
```
Oops this is a different kind of testing, this file should stay around for now. Even though it'll be automated soon™

plexus replied:
Renaming it might make more sense or putting it in a "chestnut development" section or something

plexus commented on src/leiningen/new/chestnut.clj:
```diff
@@ -20,7 +20,7 @@
 
 (def valid-options
-  ["http-kit" "site-middleware" "om-tools" "cljx" "less" "sass" "speclj"])
+  ["http-kit" "site-middleware" "om-tools" "cljx" "less" "sass" "speclj" ])
```
ah yeah sorry was gonna point this out earlier, a space snug in there

plexus commented on src/leiningen/new/chestnut.clj:
```diff
@@ -140,8 +149,13 @@
-                             "resources/public/js/polyfill.js")))
+                             "spec/cljs/chestnut/core_spec.cljs")
+          (not (speclj? opts)) (conj "env/test/js/unit-test.js"
+                                     "env/test/unit-test.html"
```
:+1:

plexus commented:
Made a few inline remarks but nothing major. Let me know if you still want to continue polishing this, otherwise I'm fine with merging it as is.

:+1: on the `env/test`, does indeed fit in nicely

I have a first working test for Chestnut itself, will open a PR in a minute so you can see what it's like. This is gonna make my life so much easier :)

plexus commented on src/leiningen/new/chestnut/project.clj:
```diff
@@ -66,20 +67,18 @@
-                   :test-commands {"spec" ["phantomjs" "bin/speclj" "resources/public/js/app_spec.js"]}{{/speclj?}}}
+                   :cljsbuild {:test-commands { {{{test-command-name}}} {{{test-command}}} }
+                               :builds {
+                                        :app {:source-paths ["env/dev/cljs"]}
```
one newline too many here

plexus commented:
Chestnut testing: https://github.com/plexus/chestnut/pull/89

rorygibson commented:
I'd removed the testing.md from ```doc/``` by accident; reverted.

Whitespace issues fixed & version name changed in the README changelog.

I also updated the ```doc/CHANGELOG.md``` to match my addition in the README, I'd missed that before.

Happy for you to merge as soon as you're OK with everything.

plexus commented:
Splendid!

plexus commented:
:shipit: 

![ship](https://cloud.githubusercontent.com/assets/32212/5977389/d9587244-a897-11e4-8ea0-0675799c7d2f.gif)

plexus commented:
> I also updated the doc/CHANGELOG.md to match my addition in the README, I'd missed that before.

That duplication should probably go away at some point... Right now I sync from README to docs when releasing.
